pkg/image/registryclient/client_mirrored.go

```diff
@@ -0,0 +1,477 @@
+	// OnFailure is invoked if FirstRequest returned no error and a nil list of locations if and only if
+	// an API call fails on the specified request. The result of alternateRepositories is cached for
+	// subsequent calls to that repository.
+	OnFailure(ctx context.Context, locator reference.DockerImageReference) (alternateRepositories []reference.DockerImageReference, err error)
+}
+
+// mirroredRepoRetriever allows a caller to retrieve a distribution.Repository. It may perform
+// requests to authorize the client and will return an error if it fails.
+type mirroredRepoRetriever interface {
+	connectToRegistry(context.Context, repositoryLocator, bool) (distribution.Repository, error)
+}
+
+// repositoryLocator caches the components necessary to connect to a single image repository.
+type repositoryLocator struct {
+	ref reference.DockerImageReference
+	// url may specify a default protocol (http) instead of (https), but is otherwise calculated
+	// by taking ref.Registry and applying it to url.Host
+	url *url.URL
+	// named is the image repository path on the server (namespace and name in ref terms) and is
+	// required for the distribution registryclient.
+	named distributionreference.Named
+}
+
+// repository provides serial lookup behavior for a given repository on errors by delegating
+// to the provided strategy for the first request or when a failure occurs.
+type mirroredRepository struct {
+	locator  repositoryLocator
+	insecure bool
+
+	strategy  AlternateBlobSourceStrategy
+	retriever mirroredRepoRetriever
+
+	lock  sync.Mutex
+	order []reference.DockerImageReference
+	repos map[reference.DockerImageReference]distribution.Repository
+}
+
+// Named returns the name of the repository.
+func (r *mirroredRepository) Named() distributionreference.Named {
+	return r.locator.named
+}
+
+// Manifests wraps the manifest service in a mirroredManifest for shared retries.
+func (r *mirroredRepository) Manifests(ctx context.Context, options ...distribution.ManifestServiceOption) (distribution.ManifestService, error) {
+	return &mirroredManifest{repo: r, options: options}, nil
+}
+
+// Blobs wraps the blob service in a mirroredBlobstore for shared retries.
+func (r *mirroredRepository) Blobs(ctx context.Context) distribution.BlobStore {
+	return mirroredBlobstore{repo: r}
+}
+
+// Tags lists the tags under the named repository.
+func (r *mirroredRepository) Tags(ctx context.Context) distribution.TagService {
+	return mirroredTags{repo: r}
+}
+
+var (
+	errNoValidAlternates = fmt.Errorf("no valid alterative sources for this content located")
+	errNoValidSource     = fmt.Errorf("no source repository defined for accessing the repository")
+)
+
+// initialRepos returns a list of locations to attempt to access, a boolean indicating that alternates
+// were suggested, or an error.
+func (r *mirroredRepository) initialRepos(ctx context.Context) ([]reference.DockerImageReference, bool, error) {
```

Comment:
You can safely ditch this bool without sacrificing performance/readability.

Reply:
It's subtle and necessary - the presence of the bool tells us whether the FirstRequest() gave us alternates or not.  If it did, we never call OnFailure().  If it doesn't, we have to call OnFailure later.  The difference is because the returned list of alternates always contains the source in the default path, so an empty alternates list (or nil) doesn't tell us what expected outcome from FirstRequest was().